Kapp2termcons

radumereuta opened the pull request:
This is supposed to be a fix for #985
I've added a new filter at parsing time that transforms instances of KApp into the equivalent TermCons
This way the type inferencer can apply correctly

I also found a pretty-printing issue in TreeNodesToKORE.scala and simplified the code a bit.

radumereuta committed:
Fix TermCons.toString for bracket productions

radumereuta committed:
Adding KApp to TermCons filter

Needs tests

radumereuta committed:
Work on transformer

and adding unit test

radumereuta committed:
Merge branch 'master' into kapp2termcons

radumereuta committed:
Fix a failing test.

radumereuta commented on kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java:
```diff
@@ -0,0 +1,99 @@
+        assert tc.production() != null : this.getClass() + ":" + " production not found." + tc;
+        if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KApply")) {
+            if (!(tc.items().get(1) instanceof Constant) || !((Constant) tc.items().get(1)).production().sort().equals(Sorts.KLabel()))
+                // TODO: maybe return a hidden warning?
```
Not sure about adding a warning here.

dwightguth commented on kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java:
```diff
@@ -0,0 +1,99 @@
+    public Either<java.util.Set<KEMException>, Term> apply(TermCons tc) {
+        assert tc.production() != null : this.getClass() + ":" + " production not found." + tc;
+        if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KApply")) {
+            if (!(tc.items().get(1) instanceof Constant) || !((Constant) tc.items().get(1)).production().sort().equals(Sorts.KLabel()))
```
Let's use TermCons.get everywhere so we don't get confused by the reversed indices

dwightguth commented on kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java:
```diff
@@ -0,0 +1,99 @@
+    /**  Recurse under #KList and flatten all the terms */
+    private static PStack<Term> flattenKList(Term t) {
+        if (t instanceof Ambiguity) {
+            Ambiguity amb = (Ambiguity) t;
```
The logic in this block doesn't look sound to me. You are basically dropping some of the parses in the forest

radumereuta replied:
I am applying an early `prefer` filter, only for `#KList`. If I get a KApp, I gather all the KList items and put them in a single list. If I get different ambiguities, like user lists, I discard them since I'm only interested in KList in this place.
I want to apply this filter in exactly this place because it simplifies this part. It is possible to have user lists mixed in, then the number of KList items is going to be smaller. That's why I'm only taking the list with the highest amount of elements. Since KList has the left attribute, I'm only going to find one element with the highest amount of items.

It's possible that I have a production with arity 2, but a KList with 3 elements. Then that's going to be an error. You can still write that if you add a set of braces (see test:495)

If I make it the first filter, I have to handle more cases when flattening KList. It could also hinder priorities. I can write `mul(plus(1,2),3)` but when I transform it into TermCons, it's going to be rejected by the priority filter.

dwightguth commented on kore/src/main/scala/org/kframework/parser/TreeNodesToKORE.scala:
```diff
@@ -29,21 +29,23 @@ class TreeNodesToKORE(parseSort: java.util.function.Function[String, Sort], stri
+  def termConsToKApply(t: TermCons): K = {
+    if (t.production.att.contains("recordPrd", classOf[Production])) {
+      val realProd = t.production.att.get("recordPrd", classOf[Production])
+      val map = t.items.asScala.reverse.zipWithIndex.map { case (item, idx) => (t.production.nonterminal(idx).name.get, apply(item))} toMap
```
This needs to be converted to an arraylist still because otherwise reverse will take quadratic time.

dwightguth commented on kore/src/main/scala/org/kframework/parser/TreeNodesToKORE.scala:
```diff
@@ -29,21 +29,23 @@ class TreeNodesToKORE(parseSort: java.util.function.Function[String, Sort], stri
+      if (t.production.klabel.isEmpty)
+        throw KEMException.internalError("Missing klabel in production: " + t.production, t)
+      val klabel = if (t.production.klabel.get.name == "#OuterCast") KLabel("project:" ++ t.production.sort.toString) else t.production.klabel.get
+      KApply(klabel.head, KList(t.items.asScala.reverse map apply asJava), locationToAtt(t.location, t.source).add(classOf[Production], realProd))
```
Same here

dwightguth commented:
It looks alright aside from the minor issues I mentioned, but it really should be the first step in the disambiguation pipeline, not left until after priority is done.

radumereuta committed:
Fix PR issues.

radumereuta committed:
Merge branch 'master' into kapp2termcons

radumereuta commented:
I've fixed most of the comments and answered to another.
I would also want you to look at the `TODO` I've added. I'm not sure what to do if I fail to transform the KApp into TermCons (cases like a KApp with a variable instead of KLabel constant). One suggestion was to add a hidden warning, but I'm not sure what is the support or future plans for that kind of pattern.

radumereuta committed:
Accept multiple possible parses under KApp

radumereuta committed:
Merge branch 'master' into kapp2termcons

dwightguth commented on kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java:
```diff
@@ -43,57 +51,68 @@ public KAppToTermConsVisitor(Module mod) {
+                Set<Term> rez = Sets.newHashSet();
+                for (Term t00 : t0)
+                    for (Term t11 : t1)
+                        rez.add(TermCons.apply(ConsPStack.singleton(t00).plus(t11), tc.production(), tc.location(), tc.source()));
```
Pretty sure this is reversed from the order you want.

dwightguth replied:
This is wrong, I misunderstood that plus does a prepend in constant time rather than an append.

dwightguth commented on kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java:
```diff
@@ -43,57 +51,68 @@ public KAppToTermConsVisitor(Module mod) {
-                return flattenKList(tc.items().get(1)).plusAll(flattenKList(tc.items().get(0)));
-            else if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#EmptyKList"))
+            if (tc.production().klabel().isDefined() && tc.production().klabel().get().equals(KLabels.KLIST))
+                return flattenKList(tc.get(0)).plusAll(flattenKList(tc.get(1)));
```
Pretty sure this is reversed too

dwightguth replied:
I was wrong here too; plusAll prepends the argument in reverse order, so if I have ((1, 2),(3, 4)) as the children, I am going to end up with 1,2,4,3 as the final arguments.

dwightguth replied:
the second flattenKList call needs to be reversed.

dwightguth reviewed (changes requested):
This needs some tests that will actually break if the children of a TermCons get reordered.

radumereuta commented on kernel/src/test/java/org/kframework/parser/concrete2kore/RuleGrammarTest.java:
```diff
@@ -472,4 +474,35 @@ public void testLayout() {
+                "  syntax Lst2 ::= E \",\" Lst2 [klabel(constr), symbol]\n" +
+                "                | Lst \n" +
+                "endmodule";
+        parseRule("constr(I, L) => L", def, 0,
```
Test 1

radumereuta commented on kernel/src/test/java/org/kframework/parser/concrete2kore/RuleGrammarTest.java:
```diff
@@ -472,4 +474,35 @@ public void testLayout() {
+                                KApply(KLabel("#SemanticCastToLst2"), KToken("L",Sort("#KVariable")))),
+                        KApply(KLabel("#SemanticCastToLst2"), KToken("L",Sort("#KVariable")))
+                )));
+        parseRule("`constr`(I, (a, L)) => L", def, 0,
```
Test 2

radumereuta committed:
Fix reverse order and extend test.

radumereuta committed:
Merge branch 'master' into kapp2termcons

radumereuta committed:
Merge branch 'master' into kapp2termcons
